mesonbuild/dependencies/blas_lapack.py

```diff
@@ -0,0 +1,136 @@
+#       MKL has separate .pc files for this (_seq vs. _iomp), OpenBLAS does not
+#
+# TODO: what about blas/lapack vs. cblas/clapack?
+#       See http://nicolas.limare.net/pro/notes/2014/10/31_cblas_clapack_lapacke/ for details
+#
+# NOTE: we ignore static libraries for now
+
+
+@factory_methods({DependencyMethods.PKGCONFIG, DependencyMethods.CMAKE, DependencyMethods.SYSTEM})
+def openblas_factory(env: 'Environment', for_machine: 'MachineChoice',
+                  kwargs: T.Dict[str, T.Any],
+                  methods: T.List[DependencyMethods]) -> T.List['DependencyGenerator']:
+    candidates: T.List['DependencyGenerator'] = []
+
+    if DependencyMethods.CMAKE in methods:
+        candidates.append(functools.partial(
+            CMakeDependency, 'OpenBLAS', env, kwargs))
+
+    if DependencyMethods.SYSTEM in methods:
+        candidates.append(functools.partial(OpenBLASDependencySystem, 'openblas', env, kwargs))
+
+    return candidates
+
+
+class OpenBLASDependencySystem(SystemDependency):
+    def __init__(self, name: str, environment: 'Environment', kwargs: T.Dict[str, T.Any]) -> None:
+        super().__init__(name, environment, kwargs)
+        self.feature_since = ('0.64.0', '')
+
+        self.env = environment
+        self.name = 'openblas'
+        self.is_found = False
+
+        # First, look for paths specified in a machine file
+        props = self.env.properties[self.for_machine].properties
+        if any(x in props for x in ['openblas_includedir', 'openblas_librarydir']):
+            self.detect_openblas_machine_file(props)
+
+        # Then look in standard directories by attempting to link
+        if not self.is_found:
+            extra_dirs = []  # TODO - for Windows may be, e.g.: Path('C:/opt/openblas/if_32/64/')
+            link_arg = self.clib_compiler.find_library('openblas', self.env, extra_dirs)
+            h = self.clib_compiler.has_header('openblas_config.h', '', self.env, dependencies=[self])
+            if link_arg and h[0]:
+                self.is_found = True
+                self.link_args += link_arg
+
+        if self.is_found:
+            self.version = self.detect_openblas_version()
+
+        self.run_check()
+
+    def detect_openblas_machine_file(self, props: dict) -> None:
+        incdir = props.get('openblas_includedir')
+        assert incdir is None or isinstance(incdir, str)
+        libdir = props.get('openblas_librarydir')
+        assert libdir is None or isinstance(libdir, str)
+
+        if incdir and libdir:
+            self.is_found = True
+            inc_dir = Path(incdir)
+            lib_dir = Path(libdir)
+            if not inc_dir.is_absolute() or not lib_dir.is_absolute():
+                # TODO: why doesn't this fail the build? It gets caught, how to avoid?
+                raise DependencyException('Paths given for openblas_includedir and openblas_librarydir in machine file must be absolute')
```

Comment:
DependencyException is handled by catching it, converting the error message to:

```python
f'Dependency lookup for {name} with method {c.func.log_tried()!r} failed: {e}'
```

logging it as a debug message, and then re-raising it if and only if the dependency is required.

DependencyException is thus meant to be raised for cases where "the dependency cannot be found", but not necessarily for cases where "your machine file is invalid what did you even do". Some other dependency, perhaps even just a base mesonlib.MesonException, is the best thing to raise if you want it to always fail the build and complain that the configuration is "flat-out wrong".

Reply:
> Some other dependency, perhaps even just a base mesonlib.MesonException, is the best thing to raise if you want it to always fail the build and complain that the configuration is "flat-out wrong".

Thanks, that is what I wanted to achieve indeed. If we can detect that config file content is wrong, I'd much prefer an exception over silently ignoring the problem (it gets logged in `meson-log.txt`, but that's too silent for the average user).